core/trino-main/src/main/java/io/trino/json/JsonPathInvocationContext.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.json;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import io.trino.json.ir.IrJsonPath;
+import io.trino.metadata.FunctionManager;
+import io.trino.metadata.Metadata;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.type.TypeManager;
+
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A class representing state, used by JSON-processing functions: JSON_EXISTS, JSON_VALUE, and JSON_QUERY.
+ * It is instantiated per driver, and allows gathering and reusing information across the processed rows.
+ * Currently, it carries two kinds of information:
+ * - JSON path, which is constant for every processed row
+ * - cached ResolvedFunctions used by certain path nodes
+ */
+public class JsonPathInvocationContext
+{
+    private final Metadata metadata;
+    private final TypeManager typeManager;
+    private final FunctionManager functionManager;
+
+    private JsonPathEvaluator evaluator;
+
+    public JsonPathInvocationContext(Metadata metadata, TypeManager typeManager, FunctionManager functionManager)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+        this.functionManager = requireNonNull(functionManager, "functionManager is null");
+    }
+
+    /**
+     * Each row evaluates the same JSON path using different set of parameters.
+     * The evaluator is initialized at the first processed row, and then reused for subsequent rows.
+     * It allows to cache resolved functions used by certain JSON path nodes, and reuse them across
+     * rows, instead of resolving them for each row separately. Caching is realised by the CachingResolver,
+     * and it addresses the assumption that all or most rows shall provide values of the same types to
+     * certain JSON path operators.
+     */
+    public void initializeIfNecessary(IrJsonPath path, ConnectorSession session)
+    {
+        if (evaluator != null) {
```

Comment:
I think that `JsonPathInvocationContext` is needed. It is passed through the `instanceFactory` and this way an instance of `JsonPathInvocationContext` persists through multiple rows processed by the JSON-funciton.

We can't skip it and just pass `JsonPathEvaluator` as the context object, because we can create the `JsonPathEvaluator` only after the function is invoked with certain arguments (including path). The context object already exists at that moment (it comes as one of the arguments to the function).

Reply:
Yes, you're right. This is the state object for the (stateful) function. However, it could be changed to be just that -- a state object, and leave all logic to be done in the caller.